Modules/_zstd/_zstdmodule.c

```diff
@@ -0,0 +1,968 @@
+            Custom dictionary content.
+        samples_bytes: PyBytesObject
+            Concatenation of samples.
+        samples_size_list: object(subclass_of='&PyList_Type')
+            List of samples' sizes.
+        dict_size: Py_ssize_t
+            The size of the dictionary.
+        compression_level: int
+            Optimize for a specific zstd compression level, 0 means default.
+        /
+
+Internal function, finalize a zstd dictionary.
+[clinic start generated code]*/
+
+static PyObject *
+_zstd__finalize_dict_impl(PyObject *module, PyBytesObject *custom_dict_bytes,
+                          PyBytesObject *samples_bytes,
+                          PyObject *samples_size_list, Py_ssize_t dict_size,
+                          int compression_level)
+/*[clinic end generated code: output=9c2a7d8c845cee93 input=ef6f4f2b0b7e178d]*/
+{
+    if (ZSTD_versionNumber() < 10405) {
+        /* Must be dynamically linked */
+        PyErr_Format(PyExc_NotImplementedError,
+                "The _finalize_dict function is only available when the underlying "
+                "zstd library's version is greater than or equal to v1.4.5. "
+                "The current zstd version is v%s.",
+                ZSTD_versionString());
+        return NULL;
+    }
+
+    Py_ssize_t chunks_number;
+    size_t *chunk_sizes = NULL;
+    PyObject *dst_dict_bytes = NULL;
+    size_t zstd_ret;
+    ZDICT_params_t params;
+    Py_ssize_t sizes_sum;
+    Py_ssize_t i;
+
+    /* Check arguments */
+    if (dict_size <= 0) {
+        PyErr_SetString(PyExc_ValueError, "dict_size argument should be positive number.");
+        return NULL;
+    }
+
+    chunks_number = Py_SIZE(samples_size_list);
+    if ((size_t) chunks_number > UINT32_MAX) {
+        PyErr_SetString(PyExc_ValueError,
+                        "The number of samples should <= UINT32_MAX.");
+        return NULL;
+    }
+
+    /* Prepare chunk_sizes */
+    chunk_sizes = PyMem_Malloc(chunks_number * sizeof(size_t));
+    if (chunk_sizes == NULL) {
+        PyErr_NoMemory();
+        goto error;
+    }
+
+    sizes_sum = 0;
+    for (i = 0; i < chunks_number; i++) {
+        PyObject *size = PyList_GET_ITEM(samples_size_list, i);
+        chunk_sizes[i] = PyLong_AsSize_t(size);
+        if (chunk_sizes[i] == (size_t)-1 && PyErr_Occurred()) {
+            PyErr_SetString(PyExc_ValueError,
+                            "Items in samples_size_list should be an int "
+                            "object, with a size_t value.");
+            goto error;
+        }
+        sizes_sum += chunk_sizes[i];
+    }
+
+    if (sizes_sum != Py_SIZE(samples_bytes)) {
+        PyErr_SetString(PyExc_ValueError,
+                        "The samples size list doesn't match the concatenation's size.");
+        goto error;
+    }
+
+    /* Allocate dict buffer */
+    dst_dict_bytes = PyBytes_FromStringAndSize(NULL, dict_size);
+    if (dst_dict_bytes == NULL) {
+        goto error;
+    }
+
```

Comment:
From the beginning of the function to this line, it appears that it's the same code as in the train() routine above. Can the logic be refactored into a single function?

Reply:
I think I will leave refactoring this and the instance below to a future PR. I'll add TODOs and try to address this after b1.